tempesta_fw/sock_clnt.c

```diff
@@ -241,6 +244,27 @@ static const SsHooks tfw_sock_clnt_ss_hooks = {
 	.connection_recv	= tfw_connection_recv,
 };
 
+static int
+__cli_clonn_close_cb(TfwConnection *conn)
+{
+	/*
+	 * Use assynchronous closing to release peer connection list and
+	 * client hash bucket locks as soon as possible and let softirq
+	 * do all the jobs.
+	 */
+	return ss_close(conn->sk);
+}
+
+static int
+tfw_cli_conn_close_all(TfwClient *cli)
+{
+	TfwConnection *conn;
+
+	tfw_peer_for_each_conn(cli, conn, list, __cli_clonn_close_cb);
```

Comment:
Yes, this is somewhat tricky. Actually `tfw_peer_for_each_conn()` returns a value, so if the work queue is overflown, then `tfw_cli_conn_close_all()` returns an error and we continue our closing loop in `tfw_sock_clnt_stop_all()`. I tried to make the code generic for server and client connections, but it seems returning from `tfw_peer_for_each_conn()` makes a reader wondering. Peer locks are also an issue - it's very desired to keep locking inside peer.h. I reworked the code for explicit returning.